internal/featuretests/tcpproxy_test.go

```diff
@@ -490,4 +553,434 @@ func TestTCPProxyTLSBackend(t *testing.T) {
 		),
 		TypeUrl: clusterType,
 	})
+
+	// check that both ingress_http and ingress_https are empty
+	c.Request(routeType).Equals(&v2.DiscoveryResponse{
+		Resources: resources(t,
+			envoy.RouteConfiguration("ingress_http"),
+			envoy.RouteConfiguration("ingress_https"),
+		),
+		TypeUrl: routeType,
+	})
+}
+
+// Assert that TCPProxy + a http service can be used to expose a ingress_http
+// route on the same vhost that port ingress_https is tls passthrough + proxing.
+func TestTCPProxyAndHTTPService(t *testing.T) {
+	rh, c, done := setup(t)
+	defer done()
+
+	s1 := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "secret",
+			Namespace: "default",
+		},
+		Type: "kubernetes.io/tls",
+		Data: secretdata(CERTIFICATE, RSA_PRIVATE_KEY),
+	}
+
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "backend",
+			Namespace: s1.Namespace,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Protocol:   "TCP",
+				Port:       80,
+				TargetPort: intstr.FromInt(8080),
+			}},
+		},
+	}
+
+	hp1 := &projcontour.HTTPProxy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "simple",
+			Namespace: svc.Namespace,
+		},
+		Spec: projcontour.HTTPProxySpec{
+			VirtualHost: &projcontour.VirtualHost{
+				Fqdn: "kuard-tcp.example.com",
+				TLS: &projcontour.TLS{
+					SecretName: s1.Name,
+				},
+			},
+			Routes: []projcontour.Route{{
+				Conditions: prefixCondition("/"),
+				Services: []projcontour.Service{{
+					Name: svc.Name,
+					Port: 80,
+				}},
+			}},
+			TCPProxy: &projcontour.TCPProxy{
+				Services: []projcontour.Service{{
+					Name: svc.Name,
+					Port: 80,
+				}},
+			},
+		},
+	}
+	rh.OnAdd(s1)
+	rh.OnAdd(svc)
+	rh.OnAdd(hp1)
+
+	c.Request(listenerType).Equals(&v2.DiscoveryResponse{
+		Resources: resources(t,
+			&v2.Listener{
+				// ingress_http is present for
+				// http://kuard-tcp.example.com/ -> default/backend:80
+				Name:    "ingress_http",
+				Address: envoy.SocketAddress("0.0.0.0", 8080),
+				FilterChains: envoy.FilterChains(
+					envoy.HTTPConnectionManager("ingress_http", envoy.FileAccessLogEnvoy("/dev/stdout"), 0),
+				),
+			},
+			&v2.Listener{
+				// ingress_https is present for
+				// kuard-tcp.example.com:443 terminated at envoy then forwarded to default/backend:80
+				Name:         "ingress_https",
+				Address:      envoy.SocketAddress("0.0.0.0", 8443),
+				FilterChains: filterchaintls("kuard-tcp.example.com", s1, tcpproxy(t, "ingress_https", "default/backend/80/da39a3ee5e")),
+				ListenerFilters: envoy.ListenerFilters(
+					envoy.TLSInspector(),
+				),
+			},
+			staticListener(),
+		),
+		TypeUrl: listenerType,
+	})
+
+	// check port 80 is open and the route is a 301 upgrade.
+	c.Request(routeType).Equals(&v2.DiscoveryResponse{
+		Resources: resources(t,
+			envoy.RouteConfiguration("ingress_http",
+				envoy.VirtualHost("kuard-tcp.example.com",
+					upgradeHTTPS(routePrefix("/")),
```

Comment:
If the listener on port 80 is meant to be routing traffic to "default/backend:80" as per the listener comment, doesn't this prevent that? ie. rather than routing requests, this route config always responds with a redirect?

Reply:
my bad, this comment wasn't clear. I've rewritten it.